### PR TITLE
Patch Jedi to get completions for compiled objects in Numpy and Matplotlib

### LINKED ISSUE #2162: Code completion in the Editor doesn't work properly, unlike in the IPython console
_From [marushk...@gmail.com](https://code.google.com/u/112862186956031011891/) on 2015-02-06T08:44:55Z_

Spyder Version:  2.4.0dev
Python Version:  3.4.1
Qt Version    :  4.8.6, PyQt4 (API v2) 4.10.4 on Windows
pyflakes >=0.6.0:  0.7.3 (OK)
pep8 >=0.6      :  1.4.6 (OK)
IPython >=1.0   :  2.3.1 (OK)
zmq >=2.1.11    :  14.4.1 (OK)
pygments >=1.6  :  2.0.1 (OK)
pandas >=0.13.1 :  0.14.1 (OK)
sphinx >=0.6.6  :  1.2.3 (OK)
jedi >=0.8.1    :  0.8.1-final0 (OK)
rope >=0.9.2    :  0.9.4-1 (OK)
matplotlib >=1.0:  1.4.0 (OK)
sympy >=0.7.0   :  0.7.5 (OK)
pylint >=0.25   :  1.4.1 (OK) 

What steps will reproduce the problem?
1. Import module 
2. Create a local variable using class from the module 
3. Write the name of the variable, put the dot and call the autocompletion (TAB or CTRL+SPACE) 

What is the expected output? What do you see instead?

 In the editor:
I expected popup box with available methods and members, but it don't appear.
In the console:
All is OK. 

Please provide any additional information below

I tried numpy and cv2:

``` python
import numpy as np # 'imp' completion works, 'num' completion works
arr = np.array([1,2,3,4,5]) # 'np.' completion works
arr. # completion don't work 

import cv2
cap = cv2.VideoCapture(0) # 'cv2.' completion works
cap. # completion don't work 
```

Same code in console work fine. I use Anaconda and last spyder version from bitbucket and run it like follow:
"python bootstrap.py --debug"
When I call completion for variable arr. or cap.  I got next debug output:
completions request:
[]
completions request from jedi complete: "[]" in 0.1 sec
completions request from rope complete: "[]" in 0.2 sec
got timeout

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=2162_

### PR DESCRIPTION
**Pull request by bcolsen**
This pull request uses code from davidhalter/jedi#796 to add completions of returns that are in numpydoc doc format.  This pull fixes #2162 for numpy and matplotlib.

In order for this to work you need the "numpydoc" module installed

Here is some "test" code to copy into the Spyder editor.

    import numpy as np
    import matplotlib.pyplot as plt
    
    x = np.array([1,2,3])
    y = np.ones()
    y.                        # works
    x.                        # works
    w = x + y
    w.                        # works
    v = y.transpose()         # finds the docstring with Ctrl-i and params pop-up
    v.                        # works
    
    fig = plt.figure()
    fig.                      # works
    fig.add_axes()            # finds the docstring with Ctrl-i and params pop-up
    
    ax = plt.subplot(111)
    ax.                       # works (fails on python 2.7...that will have to wait for numpydoc in matplotlib)
    ax.set_title()            # finds the docstring with Ctrl-i and params pop-up

### CODE REVIEW

File: spyder/docstrings.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
+"""
```
Comment (ccordoba12) [inline spyder/docstrings.py]: Let's move this file to the `spyder/utils/introspection` folder

File: spyder/docstrings.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+                m = re.match('([^,]+(,[^,]+)*?)(,[ ]*optional)?$', p_type)
+                if m:
+                    p_type = m.group(1)
+                return _expand_typestr(p_type)
+        return []
+
+    def _search_return_in_numpydocstr(docstr):
+        r"""
+        Search `docstr` (in numpydoc format) for type(-s) of `param_str`.
+
```
Comment (ccordoba12) [inline spyder/docstrings.py]: Remove this blank line

File: spyder/docstrings.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    >>> _search_param_in_docstr('@type param: int', 'param')
+    ['int']
+    >>> _search_param_in_docstr(
+    ...   ':type param: :class:`threading.Thread`', 'param')
+    ['threading.Thread']
+    >>> bool(_search_param_in_docstr('no document', 'param'))
+    False
+    >>> _search_param_in_docstr(':param int param: some description', 'param')
+    ['int']
+
```
Comment (ccordoba12) [inline spyder/docstrings.py]: Remove this blank line

File: spyder/docstrings.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    >>> _strip_rst_role(':class:`ClassName`')  # strip off :class:
+    'ClassName'
+    >>> _strip_rst_role(':py:obj:`module.Object`')  # works with domain
+    'module.Object'
+    >>> _strip_rst_role('ClassName')  # do nothing when not ReST role
+    'ClassName'
+
+    See also:
+    http://sphinx-doc.org/domains.html#cross-referencing-python-objects
+
```
Comment (ccordoba12) [inline spyder/docstrings.py]: Remove this blank line

**Comment by ccordoba12** [comment]
Thanks so much for working on this! I left some minor style comments for now :-)

To fix our tests on CircleCI you need to rebase your work on top of the latest 3.x branch.

Besides, it'd be nice if you could add some tests for this new addition and also move our tests at the end of `jedi_plugin.py` inside `spyder/utils/introspection/test/test_jedi_plugin.py`. You could use as inspiration this file

https://github.com/spyder-ide/spyder/pull/3715/files#diff-4e3108d8800283797868af8e23210200

**Comment by bcolsen** [comment]
@ccordoba12 I can make all these changes, but I think I need to make slightly larger changes for the compiled objects returns. 

I'll patch Jedi like rope is currently with a function in a separate file (like jedi_patch.py) so I can patch some Jedi methods. 

I can also set a limit on the patch to Jedi == 0.9.0 because I hope all this will be included for the 0.10.0 release.

Should I make these changes in this pull request or make a new one?  I think I can do it here without making a mess.

**Comment by ccordoba12** [comment]
> I think I can do it here without making a mess.

Sure, please make them here :-)

**Comment by bcolsen** [comment]
Numpy completes in Spyder with this commit but I still need to finish the testing and something else is wrong in matplotlib.

**Comment by ccordoba12** [comment]
> Where do I see the output of the debug_print()

In the terminal where you started Spyder, but for that you need to start Spyder like this

    SPYDER_DEBUG=3 python bootstrap.py

(assuming your using our `bootstrap.py` script to start Spyder from your git clone :-).

**Comment by bcolsen** [comment]
@ccordoba12 I've had some good progress with this. I now have mpl Axes objects returning without modifying matplotlib. I edited the first comment to show what works now.

I added a numpydoc dependancy to try to fix the CircleCl test but it still doesn't pass.

I think I'm ready for review.

**Comment by goanpeca** [comment]
@bcolsen awesome. Thanks for this work.

I think you need to add numpydoc [here](https://github.com/spyder-ide/spyder/blob/master/circle.yml#L14)

This is a pain right now, so we are moving to use a requirements.txt file so that we only change stuff there and all CI pick up the right thing.

**Comment by bcolsen** [comment]
Thanks @goanpeca. I'll add numpydoc to the circle.yml

I'm going to change numpydoc to an optional dependency as well since both numpy and matplotlib are optional.  I'll change the test_jedi_plugin to make the numpy and matplotlib tests skippable using pytest.skipif if they are not installed

**Comment by ccordoba12** [comment]
I have to say you've a terrific job here, congrats!

You're tests are all passing in Python 3.5, but unfortunately failing in 2.7 (as you can see in AppVeyor and CircleCI). Please take a look at them :-)

**Comment by bcolsen** [comment]
OK green lights....2.7 that's still a thing :-) It was a unicode problem in the docstrings. Unfortunatly, there is still a problem in the matplotlib Axes return for python 2.7, but I'm pretty amazed it worked in 3.5 without docstrings. I guess I have to stop this PR somewhere.

I'll start looking at the upstream for jedi 0.10.0 and see if I can get some of these patches in there.

Now this is ready for review.

File: spyder/utils/introspection/jedi_patch.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        # ...adding docstrings int _execute_function...
+        def _execute_function(self, evaluator, params):
+            if self.type != 'funcdef':
+                return
+            # patching docstrings here
+            from spyder.utils.introspection import docstrings
+            types = docstrings.find_return_types(evaluator, self)
+            if types:
+                for result in types:
+                    debug.dbg('BC!!!!!!!!!!!!!! wow types?: %s in %s', result, self)
```
Comment (ccordoba12) [inline spyder/utils/introspection/jedi_patch.py]: Could you improve this debug message?

Comment (ccordoba12) [inline spyder/utils/introspection/jedi_patch.py]: Or maybe not :-) This comes from Jedi source code, right?

File: spyder/utils/introspection/jedi_patch.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+                debug.warning('calculate_children StopIteration %s', types)
+        debug.dbg('calculate_children types %s', types)
+        return types
+
+    jedi.evaluate.precedence.calculate_children = calculate_children
+
+    # [4] Fixing introspection for matplotlib Axes objects
+    # Patching jedi.evaluate.precedence...
+    from jedi.evaluate.representation import \
+        tree, InstanceName, Instance, compiled, FunctionExecution, InstanceElement
```
Comment (ccordoba12) [inline spyder/utils/introspection/jedi_patch.py]: Parenthesis here too

File: spyder/utils/introspection/docstrings.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+import re
+from itertools import chain
+from textwrap import dedent
+from jedi import debug
+from jedi.evaluate.cache import memoize_default
+from jedi.parser import Parser, load_grammar
+from jedi.parser.tree import Class
+from jedi.common import indent_block
+from jedi.evaluate.iterable import Array, FakeSequence, AlreadyEvaluated
+
```
Comment (goanpeca) [inline spyder/utils/introspection/docstrings.py]: Could we please arrange the imports as

```python
# Standard library imports
from ast import literal_eval
from itertools import chain
from textwrap import dedent
import re

# Third party imports
from jedi import debug
from jedi.common import indent_block
from jedi.evaluate.cache import memoize_default
from jedi.evaluate.iterable import AlreadyEvaluated, Array, FakeSequence
from jedi.parser import Parser, load_grammar
from jedi.parser.tree import Class
```

File: spyder/utils/introspection/docstrings.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+        for p_name, p_type, p_descr in params:
+            if p_name == param_str:
+                m = re.match('([^,]+(,[^,]+)*?)(,[ ]*optional)?$', p_type)
+                if m:
+                    p_type = m.group(1)
+                return _expand_typestr(p_type)
+        return []
+
+    def _search_return_in_numpydocstr(docstr):
+        r"""
```
Comment (goanpeca) [inline spyder/utils/introspection/docstrings.py]: why the r here?

File: spyder/utils/introspection/docstrings.py [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+            m = re.match('([^,]+(,[^,]+)*?)$', p_type)
+            if m:
+                p_type = m.group(1)
+            found.extend(_expand_typestr(p_type))
+        return found
+
+
+def _expand_typestr(p_type):
+    """
+    Attempts to interpret the possible types
```
Comment (goanpeca) [inline spyder/utils/introspection/docstrings.py]: This should be

`"""Attempts to interpret the possible types."""`

**Comment by ccordoba12** [comment]
I left some minor, style comments. After that, I'd say this one is ready :-)

Thanks a lot for this great contribution!

File: spyder/utils/introspection/docstrings.py [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+    """
+    match = REST_ROLE_PATTERN.match(type_str)
+    if match:
+        return match.group(1)
+    else:
+        return type_str
+
+
+def _evaluate_for_statement_string(evaluator, string, module):
+    if string is None:
```
Comment (goanpeca) [inline spyder/utils/introspection/docstrings.py]: docstring?

File: spyder/utils/introspection/docstrings.py [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+    code = dedent("""
+    def pseudo_docstring_stuff():
+        # Create a pseudo function for docstring statements.
+    %s
+    """)
+
+    for element in re.findall('((?:\w+\.)*\w+)\.', string):
+        # Try to import module part in dotted name.
+        # (e.g., 'threading' in 'threading.Thread').
+        string = 'import %s\n' % element + string
```
Comment (goanpeca) [inline spyder/utils/introspection/docstrings.py]: We prefer the use of format and like

```python

string = 'import {0}\n'.format(element) + string
```

File: spyder/utils/introspection/docstrings.py [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+
+    # Take the default grammar here, if we load the Python 2.7 grammar here, it
+    # will be impossible to use `...` (Ellipsis) as a token. Docstring types
+    # don't need to conform with the current grammar.
+    p = Parser(load_grammar(), code % indent_block(string))
```
Comment (goanpeca) [inline spyder/utils/introspection/docstrings.py]: same here

File: spyder/utils/introspection/docstrings.py [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+        # call. In that case it's the module of the function call.
+        # stuffed with content from a function call.
+        pseudo_cls.parent = module
+        type_list = _execute_types_in_stmt(evaluator, stmt)
+    return type_list
+
+
+def _execute_types_in_stmt(evaluator, stmt):
+    """
+    Executing all types or general elements that we find in a statement. This
```
Comment (goanpeca) [inline spyder/utils/introspection/docstrings.py]: First line should be a one line description then a blank line then the extra details

File: spyder/utils/introspection/docstrings.py [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+    """
+    definitions = evaluator.eval_element(stmt)
+    types_list = [_execute_array_values(evaluator, d) for d in definitions]
+    type_list = list(chain.from_iterable(types_list))
+    return type_list
+
+
+def _execute_array_values(evaluator, array):
+    """
+    Tuples indicate that there's not just one return value, but the listed
```
Comment (goanpeca) [inline spyder/utils/introspection/docstrings.py]: same here

File: spyder/utils/introspection/docstrings.py [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+    ones.  `(str, int)` means that it returns a tuple with both types.
+    """
+    if isinstance(array, Array):
+        values = []
+        for types in array.py__iter__():
+            objects = set(chain.from_iterable(_execute_array_values(evaluator, typ) for typ in types))
```
Comment (goanpeca) [inline spyder/utils/introspection/docstrings.py]: too long line?

File: spyder/utils/introspection/docstrings.py [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+            values.append(AlreadyEvaluated(objects))
+        return [FakeSequence(evaluator, values, array.type)]
+    else:
+        return evaluator.execute(array)
+
+
+@memoize_default(None, evaluator_is_first_arg=True)
+def follow_param(evaluator, param):
+    """
+    Determines a set of potential types for `param` using docstring hints
```
Comment (goanpeca) [inline spyder/utils/introspection/docstrings.py]: period at the end

File: spyder/utils/introspection/docstrings.py [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+    :type param: jedi.parser.tree.Param
+
+    :rtype: list
+    """
+    def eval_docstring(docstr):
+        param_str = str(param.name)
+        return set(
+            [p for string in _search_param_in_docstr(docstr, param_str)
+                for p in _evaluate_for_statement_string(evaluator, string, module)]
+        )
```
Comment (goanpeca) [inline spyder/utils/introspection/docstrings.py]: Isn't it better to just use two normal loops?

Using a doubly nested list comp. does not make things clear.

File: spyder/utils/introspection/docstrings.py [hunk 18]
```diff
@@ -0,0 +1,285 @@
...
+        if cls.type == 'classdef':
+            types |= eval_docstring(cls.raw_doc)
+
+    return types
+
+
+@memoize_default(None, evaluator_is_first_arg=True)
+def find_return_types(evaluator, func):
+    """
+    Determines a set of potential return types for `func` using docstring hints
```
Comment (goanpeca) [inline spyder/utils/introspection/docstrings.py]: period at the end

**Comment by ccordoba12** [comment]
@bcolsen took `instropection/docstrings.py` from an open PR in Jedi, so I think we shouldn't demand him to fix style errors in that file.

**Comment by goanpeca** [comment]
>@bcolsen took utils/instropection/docstrings.py from an open PR in Jedi, so I think we shouldn't demand him to fix style errors in that file

Ok!

File: spyder/utils/introspection/jedi_patch.py (hunk 5 above)
Comment (bcolsen) [inline spyder/utils/introspection/jedi_patch.py]: Yes...oops...I needed something easy to see. I'll make it better

**Comment by bcolsen** [comment]
I made the changes to jedi_patch.py. I really don't want to mess around with docstrings.py too much.  I hope to upstream some changes to jedi after the docstrings patch gets merged there.

Thanks for your encouragement on this PR and of coarse the great work you do on Spyder! 
3.1 is going to be a great release

**Comment by goanpeca** [comment]
Thans for this great work @bcolsen  :-)

**Comment by ccordoba12** [comment]
> 3.1 is going to be a great release

Sure it is!

**Pull request merged**